### PR TITLE
[MOB-2735] Wire Snapshots Testing to CI + improvements 

### PR DESCRIPTION
**Pull request by d4r1091**
[MOB-2735]

## Context

We want to enable snapshot testing on Ecosia iOS.
We want to wire it as part of our CI process.

## Approach

After the v1 implementation, we already foresaw improvements in its implementation.
I'll try to be as concise as possible.

> Our north ⭐ is always to pragmatically strive for the most maintainable and scalable solution.

The previous implementation required the Engineers to declare languages and devices as part of the tests.
While this was already a valuable solution to make a wrapper of the [Snapshot Testing Library](https://github.com/pointfreeco/swift-snapshot-testing), it still required modifying the test codebase whenever a new language or device would be added.
We wanted to decouple that as well as make sure each screenshot is taken against a precise device for a precise OS version (previously we were only passing traits, not running the simulator itself).
This PR serves as a reviewed implementation of the entire test declaration structure.
This PR introduces a manifest-based Snapshot Testing architecture, where the `snapshot_configuration.json` serves as the source of truth for what's being tested.

### To acknowledge

The Snapshot Testing is a quite consuming task that shouldn't run on every merge into `main`.
What this PR introduces is a separate workflow for Snapshot Testing whenever a PR containing a change into the `MARKETING_VERSION` (aka version bump) is made.
As a version bump doesn't happen as often as emerging a feature/bug in `main` and the need to QA that, this is likely the best place to put that as we do not want to bump the version before verifying that all the Snapshot Tests we made pass.

#### Tradeoffs

The majority of the past few days have been invested in finding a way to make the 1:1 comparison always with maximum precision. However, keeping the "pragmatic" approach of our North ⭐ message, the decision was not to overcomplicate the tests where a Locale different from the current one is being set. 

> By "testing different languages" , we are testing against the different strings only. There is no known way (against our Codebase) to override the `Locale.current` at runtime, so Locale dependent logic (e.g. `NumberFormatter`) will follow the Locale at the time of testing. To make tests succeed, a different `precision` parameter for these tests has been provided.
 
### Next Steps

- We still want to move those Snapshots elsewhere, likely a separate repo to pull at convenience
- Once this structure gets a few trials in the day-to-day development and releases flow, would be nice to open-source the wrapper as a separate Swift Package to get contributions and sparking overall interest in the community.

## Other

- Includes the Name of the PR Agent
- Includes an update to the "Marketing Version Check" to make it available to both CirleCI and GitHub Action.

### Checklist

- [x] I performed some relevant testing on a real device and/or simulator
- [x] I wrote Unit Tests that confirm the expected behaviour
- [x] I added the `// Ecosia:` helper comments where needed
- [x] I included documentation updates to the coding standards or Confluence doc, when needed

[MOB-2735]: https://ecosia.atlassian.net/browse/MOB-2735?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Client.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -19812,10 +19824,13 @@
...
 				CLANG_WARN_UNREACHABLE_CODE = YES;
 				CLANG_WARN__DUPLICATE_METHOD_MATCH = YES;
-				CODE_SIGN_STYLE = Automatic;
+				CODE_SIGN_ENTITLEMENTS = "";
+				CODE_SIGN_IDENTITY = "";
+				CODE_SIGN_STYLE = Manual;
 				COPY_PHASE_STRIP = NO;
 				CURRENT_PROJECT_VERSION = 1;
 				DEBUG_INFORMATION_FORMAT = dwarf;
+				DEVELOPMENT_TEAM = 33YMRSYD2L;
```
Comment (lucaschifino) [inline Client.xcodeproj/project.pbxproj:19833]: This was moved to be handled by the config files on https://github.com/ecosia/ios-browser/pull/738... any reason to add it to the project here now?

File: Client.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -19812,10 +19824,13 @@
 				CLANG_WARN_UNGUARDED_AVAILABILITY = YES_AGGRESSIVE;
 				CLANG_WARN_UNREACHABLE_CODE = YES;
 				CLANG_WARN__DUPLICATE_METHOD_MATCH = YES;
-				CODE_SIGN_STYLE = Automatic;
+				CODE_SIGN_ENTITLEMENTS = "";
+				CODE_SIGN_IDENTITY = "";
+				CODE_SIGN_STYLE = Manual;
```
Comment (lucaschifino) [inline Client.xcodeproj/project.pbxproj:19829]: Any reason for changing this now? I guess manual is indeed safer but this is not related to Snapshot Testing is it?

`PROVISIONING_PROFILE_SPECIFIER = "";` below also seems a bit strange 🤔

File: EcosiaTests/SnapshotTests/EcosiaSnapshotTests.xctestplan [hunk 3]
```diff
@@ -15,7 +15,8 @@
     {
       "parallelizable" : true,
       "skippedTests" : [
-        "NTPTests"
+        "NTPTests",
+        "SnapshotBaseTests"
```
Comment (lucaschifino) [inline EcosiaTests/SnapshotTests/EcosiaSnapshotTests.xctestplan:19]: Why do we need to skip `SnapshotBaseTests`?

File: EcosiaTests/SnapshotTests/snapshot_configuration.json
Comment (lucaschifino) [inline EcosiaTests/SnapshotTests/snapshot_configuration.json:1]: Great idea with these configs 👏

File: EcosiaTests/SnapshotTests/SnapshotTestHelper.swift [hunk 4]
```diff
@@ -107,38 +101,35 @@ final class SnapshotTestHelper {
...
-    ///   - precision: The precision of the snapshot comparison.
+    ///   - precision: The precision of the snapshot comparison. Default to 0.99 to allow slightly different colors between CI and local runs.
     ///   - file: The file in which failures should be reported.
     ///   - testName: The name of the test.
     ///   - line: The line number to report failures.
     static func assertSnapshot(
         initializingWith initializer: @escaping () -> UIViewController,
-        devices: [DeviceType] = [.iPhone12Pro_Portrait],
-        locales: [Locale] = [Locale(identifier: "en")],
+        locales: [Locale] = LocaleRetriever.getLocales(),
```
Comment (lucaschifino) [inline EcosiaTests/SnapshotTests/SnapshotTestHelper.swift:124]: I guess there is no reason to keep the `locales` parameter anymore, right? We can directly use `LocaleRetriever.getLocales()` inside `performSnapshot()`

File: SNAPSHOT_TESTING_WIKI.md
Comment (lucaschifino) [inline SNAPSHOT_TESTING_WIKI.md:1]: 😍

**Review by lucaschifino — COMMENTED** [review]
Great job, the improvements look great 👏 

I added some comments to consider before approving, but nothing major 🙂

File: .github/workflows/merge_tests.yml [hunk 5]
```diff
@@ -1,16 +1,13 @@
 name: Merge Unit Tests
 
 on:
-  pull_request:
+  push:
```
[the remark this answers is not in the document]
Comment (d4r1091) [reply .github/workflows/merge_tests.yml:4]: Yeah 💯 - that was a refuse. Will update thanks for checking 🤚

File: Client.xcodeproj/project.pbxproj (hunk 1 above)
Comment (d4r1091) [reply Client.xcodeproj/project.pbxproj:19833]: 💯 ! The config file works for all the client-facing targets. For the `EcosiaSnapshotTests` I thought about not assigning any specific config so that that config wouldn't interfere with tests. This decision led to assign the dev team for this target.

File: Client.xcodeproj/project.pbxproj (hunk 2 above)
Comment (d4r1091) [reply Client.xcodeproj/project.pbxproj:19829]: These are all part of the `EcosiaSnapshotTests` specific target. 
We do run the tests passing those values specifically at CI level as well, but if we do not update those for the target specifically, we will have issues on running the tests locally every time.

File: EcosiaTests/SnapshotTests/EcosiaSnapshotTests.xctestplan (hunk 3 above)
Comment (d4r1091) [reply EcosiaTests/SnapshotTests/EcosiaSnapshotTests.xctestplan:19]: `SnapshotBaseTests` is the base class from which all tests inherit. No tests are defined there so no need to run it and have it in our test results 👍

File: EcosiaTests/SnapshotTests/snapshot_configuration.json
Comment (d4r1091) [reply EcosiaTests/SnapshotTests/snapshot_configuration.json:1]: Thanks! Everything is managed at script level which is the component doing the heavy lifting here 😅

File: SNAPSHOT_TESTING_WIKI.md
Comment (d4r1091) [reply SNAPSHOT_TESTING_WIKI.md:1]: 😇 💪

**Review by lucaschifino — APPROVED** [review]
🧪 📸 🤖 🎉
